### ISSUE TITLE
FEAT: Adding check-pointing capabilities to PMDK

### ISSUE THREAD

**Issue by eberroca** [pmem/issues#1056]
# FEAT: Adding check-pointing capabilities to PMDK (libpmemobj, libpmemblk, libpmemlog)

## Rationale

Although it is true that persistence does not necessarily means endurance, it would be nice to have a mechanism in PMDK (libpmemobj, libpmemblk, libpmemlog) by which pools can be check-pointed to another remote location for fault tolerance reasons. Currently, there is only 2 ways of doing this and both are not 100% satisfactory.

(1) Use pool sets and replicate all writes: This option is not desirable since, even though it makes sure that all replicas are synchronized, the performance of writes is that of the lower performer. In the case of remote writing, that is the speed of the network. Do we really want to write to PMEM at the same speed--or worst--than that of flash SSDs?

(2) Stop the program and manually copy the pool: What if I don't want to stop my program? What if it takes a long while to copy my pool? I would need to wait and time is $$$

## Description

The idea is to have a mechanism in PMDK by which the pool is saved (check-pointed) to a location indicated by the application. This check-pointing can be done in 3 ways: 

(1) one-time only (just checkpoint now and forget about it). 

(2) with a predefined frequency (i.e., checkpoint every X number of seconds). 

(3) with a predefined frequency but only when the app is in a "mostly-reads phase". An algorithm can be implemented to detect these phases so as to have as minimum impact in the application's write performance as possible.

Ideally, these check-points would be incremental. Instead of copying, let say, a 1 TB pool in every checkpoint, only the latest changes would be copied. However, I am not sure how possible is this with MAP_SYNC.

Maybe this can be done in a coarse-grain way. For example, copying just blocks in the file changed since last time (not worrying about logging fine-grain changes and so forth). Maybe a bitmap needs to be added to the pool's metadata?. This check-pointing can even be done in parallel to the app. Dirty blocks can be copied in a first phase, and then stopping the application briefly only in the second phase to copy the modified blocks by the app while the 1st phase copying was taking place. Is this feasible? 

## API Changes

3 new calls would be added. Something like this:

RECURRENT:
pmemobj_checkpoint_recurrent_init ( PMEMobjpool *pop, destination, type_of_checkpoint, freq, ...)
pmemobj_checkpoint_recurrent_end ( PMEMobjpool *pop, destination, type_of_checkpoint)

ONE TIME CHECKPOINT:
pmemobj_checkpoint ( PMEMobjpool *pop, destination)

**Comment by kilobyte** [comment]
Note that this is possible with btrfs (although it's DAX support is only in an early patchset that's making rounds on kernel mailinglists) — we may want to draw some inspiration.

It can do so in two ways:
* in the mis-named "nodatacow" mode (usually set via chattr +C) which is really "cow once": if you write to a page that's shared with other snapshots/reflinks, the page fault will copy and unshare that page to a new extent.  You can then use all the usual btrfs tools to enumerate and/or send the changes between the two snapshots.
* regular mode: it CoWs once per page per filesystem commit (typically on the order of 30 seconds).  Unlike traditional disks, I consider this mode pointless on DAX unless used to implement some sort of RAID > 0 (which is not yet a part of the patchset), and bad for performance.  In this mode, the usual btrfs guarantee of every extent's generation number providing a not-after timestamp holds.

The kernel has an easier time as it knows about all dirty pages, rather than just ones that were modified via an explicit call.
